src/JwksHandler.ts

```diff
@@ -11,110 +11,125 @@ import { ServicesEnum } from "./models/enums/ServicesEnum";
+                body.keys.push(jwk);
+            } else logger.warn({ message:"Environment contains missing keys" });
+        });
 
-		const jwks = await Promise.all(
-			kmsKeyIds.map(async id => getAsJwk(id)),
-		);
-		jwks.forEach(jwk => {
-			if (jwk != null) {
-				body.keys.push(jwk);
-			} else logger.warn({ message:"Environment contains missing keys" });
-		});
+        const uploadParams = {
+            Bucket: JWKS_BUCKET_NAME,
+        Key: ".well-known/jwks.json",
+        Body: JSON.stringify(body),
+        ContentType: "application/json",
+        };
 
-		const uploadParams = {
-			Bucket: environmentVariables.jwksBucketName(),
-			Key: ".well-known/jwks.json",
-			Body: JSON.stringify(body),
-			ContentType: "application/json",
+		const copyParams = {
+			Bucket: PUBLISHED_KEYS_BUCKET_NAME,
+			Key: "jwks.json",
+			CopySource: `${JWKS_BUCKET_NAME}/.well-known/jwks.json`,
 		};
 
+        try {
+            logger.info({ message: "Uploading keys to JWKS bucket" });
+            await this.s3Client.send(new PutObjectCommand(uploadParams));
+
+            logger.info({ message: "Copying keys to published keys bucket" });
+            await this.s3Client.send(new CopyObjectCommand(copyParams));
+			logger.info({ message: "Keys copied to published keys bucket" });
+
+        return JSON.stringify(body);
+        } catch (err) {
+            logger.error({ message: "Error writing / copying keys", err });
+            throw new Error("Error writing keys");
+        }
+
+    }
+
+	async getAsJwk (keyId: string): Promise<Jwk | null> {
+		let kmsKey;
 		try {
-			await s3Client.send(new PutObjectCommand(uploadParams));
-		} catch (err) {
-			logger.error({ message: "Error writing keys to S3 bucket" + err });
-			throw new Error("Error writing keys to S3 bucket");
+			kmsKey = await this.kmsClient.getPublicKey({ KeyId: keyId });
+		} catch (error) {
+			logger.warn({ message:"Failed to fetch key from KMS" }, { error });
 		}
-		return JSON.stringify(body);
+	
+		const map = this.getKeySpecMap(kmsKey?.KeySpec);
+		if (
+			kmsKey != null &&
+			map != null &&
+			kmsKey.KeyId != null &&
+			kmsKey.PublicKey != null
+		) {
+			const use = kmsKey.KeyUsage === "ENCRYPT_DECRYPT" ? "enc" : "sig";
+			const publicKey = crypto
+				.createPublicKey({
+					key: kmsKey.PublicKey as Buffer,
+					type: "spki",
+					format: "der",
+				})
+				.export({ format: "jwk" });
+			return {
+				...publicKey,
+				use,
+				kid: keyId.split("/").pop(),
+				alg: map.algorithm,
+			} as unknown as Jwk;
+		}
+		logger.error({ message: "Failed to build JWK from key" + keyId }, JSON.stringify(map));
+		return null;
+	};
 
-	}
+	getKeySpecMap (
+		spec?: string,
+	): { keySpec: string; algorithm: Algorithm } | undefined {
+		if (spec == null) return undefined;
+		const conversions = [
+			{
+				keySpec: "ECC_NIST_P256",
+				algorithm: "ES256" as Algorithm,
+			},
+			{
+				keySpec: "RSA_2048",
+				algorithm: "RS256" as Algorithm,
```

Comment:
Do we definitely not want to fix the bug we have around using the wrong algorithm digest here? 

See [JWA for Encryption](https://www.rfc-editor.org/rfc/rfc7518.html#section-4.1):

```suggestion
				algorithm: "RSA_OAEP_256" as Algorithm,
```